File: day19.nim
import strutils, tables, sets

const instructions = readFile("./inputs/19.txt").splitLines()

var
  replacements = initTable[string, seq[string]]()
  combinations = initSet[string]()


for i in 0 .. instructions.high()-2:
  let values = instructions[i].split(" => ")
  if not replacements.hasKey(values[0]):
    replacements[values[0]] = @[values[1]]
  else:
    replacements[values[0]].add(values[1])

let molecule = instructions[^1]


for key in replacements.keys:
  let n = len(key)
  var pos = molecule.find(key)

  while pos != -1:
    for val in replacements[key]:
      var m = molecule
      m[pos ..< pos+n] = val
      combinations.incl(m)
    pos = molecule.find(key, pos+n)

echo len(combinations)

# https://www.reddit.com/r/adventofcode/comments/3xflz8/day_19_solutions/cy4etju/
var
  noOfElements: int
for c in molecule:
  if c.isUpperAscii: inc noOfElements
let
  noOfRn = molecule.count("Rn")
  noOfY = molecule.count('Y')

echo(noOfElements - 2*noOfRn - 2*noOfY - 1)

File: python/day19.py
from collections import defaultdict


def generate_replacements(data):
    result = defaultdict(list)
    for line in data[:-2]:
        k, v = line.split(" => ")
        result[k].append(v)
    return result

def part_1(molecule, replacements):
    result = set()
    for k, vals in replacements.items():
        pos = molecule.find(k)
        n = len(k)
        while pos != -1:
            for val in vals:
                sub = molecule[:pos] + val + molecule[pos+n:]
                result.add(sub)
            pos = molecule.find(k, pos+n)
    return len(result)

def part_2(molecule):
    # https://www.reddit.com/r/adventofcode/comments/3xflz8/day_19_solutions/cy4etju/
    elements = sum(c.isupper() for c in molecule)
    rn = molecule.count("Rn")
    y = molecule.count("Y")
    return elements - 2*rn - 2*y - 1


data = open("inputs/19.txt").read().splitlines()
molecule = data[-1]
replacements = generate_replacements(data)

print(part_1(molecule, replacements))
print(part_2(molecule))
